R/tree-map-alignment.R

```diff
@@ -455,16 +460,344 @@ find_best_shift_icp = function(pred, obs) {
+    # Storing the normalized min distance and the angle to the pre-allocated
+    # feature descriptor matrix
+    feat_desc_mat[, i_quadrant] = min_dist_i_norm
+    feat_desc_mat[, i_quadrant + 4] = angle_quad_i_norm
+  }
+
+  char_theta = atan2(char_dirs[, 1], char_dirs[, 2])
+  return(list(feat_desc_mat, char_theta))
+}
+
+# function [from_idx2_to_closest_idx1, from_idx2_to_min_dist1, n_of_matches] ...
+get_closest_pairs_after_transformation = function(xy_mat1, xy_mat2, R_mat, t_vect, r_thres) {
+  # Helper function for obtaining a mapping from each object in dataset 2 to the
+  # index of the closest object in dataset 1 after transforming the object
+  # locations in dataset 1 as R_mat*xy_mat1' + t_vect.
+  # The function also evaluates the number of matching pairs by finding the
+  # the number of closest object pairs, for which the pair-wise distance is below
+  # r_thres.
+  #
+  # Args:
+  #   xy_mat1 = (x,y)-coordinates of objects (e.g. trees) found from point
+  #             cloud 1 as a N1-by-2 matrix. (N1=number of rows, 2= number of
+  #             columns)
+  #   xy_mat2 = (x,y)-coordinates of objects (e.g. trees) found from point
+  #             cloud 2 as N2-by-2 matrix
+  #   R_mat = 2-by-2 rotation matrix of the transformation
+  #   t_vect = 2-by-1 column vector corresponding to the translation vector
+  #           of the transformation
+  #   r_thres = distance threshold for deciding if the closest object in the
+  #           other point cloud is a match
+  # Returns:
+  #   from_idx2_to_closest_idx1 = N2-by-1 vector containing the indices of the closest
+  #           object in dataset 1 for each object in dataset 2.
+  #   from_idx2_to_min_dist1 = N2-by-1 vector containing the distances to the closest
+  #           object in dataset 1 for each object in dataset 2.
+  #   n_of_matches = number of matching object pairs based on r_thres.
+  xy_mat1 = as.matrix(xy_mat1)
+  xy_mat2 = as.matrix(xy_mat2)
+  colnames(xy_mat2) = NULL
+  xy_mat1_t = t(xy_mat1)
+  rotated_xy1 = t(R_mat %*% xy_mat1_t)
+  t_vect = as.vector(as.numeric(t_vect))
+  xy_mat1_transformed = t(t(rotated_xy1) + t_vect) # N-by-2
+  # Pair-wise locations between each object in dataset 1 and 2. Each row is
+  # one object in dataset 2 and each column in one object in dataset 1.
+  transformed_loc_p_dist = cdist(xy_mat2, xy_mat1_transformed)
+  # Finding the closest object in transformed point cloud 1 for each object
+  # in point cloud 2
+  from_idx2_to_closest_idx1 = apply(transformed_loc_p_dist, 1, which.min)
+  from_idx2_to_min_dist1 = transformed_loc_p_dist[
+    cbind(1:nrow(transformed_loc_p_dist), from_idx2_to_closest_idx1)
+  ]
+
+  # Computing the number of matches, i.e., object pairs whose distance is
+  # below the set criterion
+  less_than_thresh = from_idx2_to_min_dist1 < r_thres
+  n_of_matches = sum(less_than_thresh)
+  return(list(from_idx2_to_closest_idx1, from_idx2_to_min_dist1, n_of_matches))
```

Comment:
It might be slightly more intuitive to make this a *named* list so that you can reference its elements by name, and future changes to the order will not break the functions using this function's result. Syntax:

`list(name = value, name2 = value2)`